modules/loader/src/test/scala/com/snowplowanalytics/snowplow/rdbloader/experimental/LoaderSpecification.scala

```diff
@@ -0,0 +1,116 @@
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and
+ * limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.experimental
+
+import scala.concurrent.duration.DurationInt
+
+import doobie.ConnectionIO
+
+import cats.effect.{IO, Resource}
+import cats.effect.std.Dispatcher
+import cats.effect.unsafe.implicits.global
+
+import fs2.concurrent.SignallingRef
+
+import org.http4s.blaze.client.BlazeClientBuilder
+
+import com.snowplowanalytics.snowplow.rdbloader.config.{Config, StorageTarget}
+import com.snowplowanalytics.snowplow.rdbloader.dsl.metrics.Metrics
+import com.snowplowanalytics.snowplow.rdbloader.dsl.{Logging, Monitoring, Transaction}
+import com.snowplowanalytics.snowplow.rdbloader.common.LoaderMessage
+import com.snowplowanalytics.snowplow.rdbloader.common.cloud.{Queue, SecretStore}
+import com.snowplowanalytics.snowplow.rdbloader.common.integrationtestutils.ItUtils._
+import com.snowplowanalytics.snowplow.rdbloader.common.integrationtestutils.InputBatch
+
+import org.specs2.mutable.Specification
+
+abstract class LoaderSpecification extends Specification with TestDAO.Provider with StorageTargetProvider with AzureTestResources {
+  skipAllIf(anyEnvironmentVariableMissing())
+  import LoaderSpecification._
+
+  def run[A](
+    inputBatches: List[InputBatch],
+    countExpectations: CountExpectations,
+    dbActions: TestDAO => IO[A]
+  ): IO[(WindowsAccumulator[WindowOutput], A)] =
+    createResources
+      .use { resources =>
+        for {
+          _ <- resources.testDAO.cleanDb
+          windowsAccumulator <- SignallingRef.of[IO, WindowsAccumulator[WindowOutput]](WindowsAccumulator(List.empty))
+          consuming = consumeAllIncomingWindows[WindowOutput](
+                        resources.queueConsumer,
+                        countExpectations,
+                        windowsAccumulator,
+                        getWindowOutput = sc => IO.pure(WindowOutput(sc))
+                      )
+          producing = produceInput(inputBatches, resources.producer)
+          _ <- consuming.concurrently(producing).compile.drain
+          collectedWindows <- windowsAccumulator.get
+          dbActionResult <- dbActions(resources.testDAO)
+        } yield (collectedWindows, dbActionResult)
+      }
+
+  def createResources: Resource[IO, TestResources] =
+    for {
+      consumer <- createConsumer
+      producer <- createProducer
+      implicit0(secretStore: SecretStore[IO]) <- createSecretStore
+      transaction <- createDbTransaction
+      testDAO = createDAO(transaction)
+    } yield TestResources(queueConsumer = consumer, producer = producer, testDAO = testDAO)
+
+  def createDbTransaction(implicit secretStore: SecretStore[IO]): Resource[IO, Transaction[IO, ConnectionIO]] = {
```

Comment:
I'm wondering whether we need `Transaction[IO, ConnectionIO]` type here in tests, with all the pooling, retries, transactions handling etc. Wouldn't simple `Transactor.fromDriverManager` from doobie be sufficient?

Reply:
I tried to use existing way in order to not duplicate the logic in the tests as well but you are right, it is quite complicated to initialize `Transaction` and we don't use most of the features in the tests. I will try to use `Transactor` directly 👍